contracts/peg/NonCustodialPSM.sol

```diff
@@ -0,0 +1,479 @@
+    {
+        require(
+            address(newMinter) != address(0),
+            "PegStabilityModule: Invalid new GlobalRateLimitedMinter"
+        );
+        GlobalRateLimitedMinter oldMinter = rateLimitedMinter;
+        rateLimitedMinter = newMinter;
+
+        emit GlobalRateLimitedMinterUpdate(oldMinter, newMinter);
+    }
+
+    /// @notice set the mint fee vs oracle price in basis point terms
+    function _setMintFee(uint256 newMintFeeBasisPoints) internal {
+        require(
+            newMintFeeBasisPoints <= MAX_FEE,
+            "PegStabilityModule: Mint fee exceeds max fee"
+        );
+        uint256 _oldMintFee = mintFeeBasisPoints;
+        mintFeeBasisPoints = newMintFeeBasisPoints;
+
+        emit MintFeeUpdate(_oldMintFee, newMintFeeBasisPoints);
+    }
+
+    /// @notice internal helper function to set the redemption fee
+    function _setRedeemFee(uint256 newRedeemFeeBasisPoints) internal {
+        require(
+            newRedeemFeeBasisPoints <= MAX_FEE,
+            "PegStabilityModule: Redeem fee exceeds max fee"
+        );
+        uint256 _oldRedeemFee = redeemFeeBasisPoints;
+        redeemFeeBasisPoints = newRedeemFeeBasisPoints;
+
+        emit RedeemFeeUpdate(_oldRedeemFee, newRedeemFeeBasisPoints);
+    }
+
+    /// @notice helper function to set the PCV deposit
+    function _setPCVDeposit(IPCVDeposit newPCVDeposit) internal {
+        require(
+            address(newPCVDeposit) != address(0),
+            "PegStabilityModule: Invalid new PCVDeposit"
+        );
+        require(
+            newPCVDeposit.balanceReportedIn() == address(underlyingToken),
+            "PegStabilityModule: Underlying token mismatch"
+        );
+        IPCVDeposit oldTarget = pcvDeposit;
+        pcvDeposit = newPCVDeposit;
+
+        emit PCVDepositUpdate(oldTarget, newPCVDeposit);
+    }
+
+    // ----------- Public State Changing API -----------
+
+    /// @notice function to redeem FEI for an underlying asset
+    /// We do not burn Fei; this allows the contract's balance of Fei to be used before the buffer is used
+    /// In practice, this helps prevent artificial cycling of mint-burn cycles and prevents a griefing vector.
+    /// This function will deplete the buffer for pulling from the PCV Deposit
+    /// @param to the destination address for proceeds
+    /// @param amountFeiIn the amount of FEI to sell
+    /// @param minAmountOut the minimum amount out otherwise the TX will fail
+    function redeem(
+        address to,
+        uint256 amountFeiIn,
+        uint256 minAmountOut
+    )
+        external
+        virtual
+        override
+        nonReentrant
+        whenNotPaused
+        whileRedemptionsNotPaused
+        returns (uint256 amountOut)
+    {
+        updateOracle();
+
+        amountOut = _getRedeemAmountOut(amountFeiIn);
+        require(
+            amountOut >= minAmountOut,
+            "PegStabilityModule: Redeem not enough out"
+        );
+        _depleteBuffer(amountFeiIn);
```

Comment:
I feel like we should rate limit in the asset that has the oracle, since this will prevent oracle manipulation attacks. Not a trivial question to answer though!

Reply:
Good catch, will remove the buffer depletion before the oracle update so that if things fail, they waste less gas.